tee-worker/litentry/core/vc-issuance/lc-vc-task-receiver/src/lib.rs

```diff
@@ -110,132 +113,195 @@ where
+		let assertion_networks = assertion.get_supported_web3networks();
+		let identities = get_eligible_identities(id_graph.as_ref(), assertion_networks);
 		ensure!(!identities.is_empty(), "No eligible identity".to_string());
 
-		let signer = signer
+		let signer_account = signer
 			.to_account_id()
 			.ok_or_else(|| "Invalid signer account, failed to convert".to_string())?;
 
+		match assertion {
+			// the signer will be checked inside A13, as we don't seem to have access to ocall_api here
+			Assertion::A13(_) => (),
+			_ => if_production_or!(
+				ensure!(ensure_self(&signer, &who), "Unauthorized signer",),
+				ensure!(
+					ensure_self(&signer, &who) || ensure_alice(&signer_account),
+					"Unauthorized signer",
+				)
+			),
+		}
+
 		let req = AssertionBuildRequest {
-			shard,
-			signer,
-			who,
-			assertion,
+			shard: request.shard,
+			signer: signer_account,
+			who: who.clone(),
+			assertion: assertion.clone(),
 			identities,
 			top_hash: H256::zero(),
 			parachain_block_number,
 			sidechain_block_number,
 			maybe_key,
+			should_create_id_graph,
 			req_ext_hash,
 		};
 
 		let vc_request_handler = VCRequestHandler { req, context: context.clone() };
-		let response = vc_request_handler
+		let res = vc_request_handler
 			.process()
 			.map_err(|e| format!("Failed to build assertion due to: {:?}", e))?;
 
 		let call_index = node_metadata_repo
 			.get_from_metadata(|m| m.vc_issued_call_indexes())
-			.unwrap()
-			.unwrap();
-		let result = aes_encrypt_default(&key, &response.vc_payload);
+			.map_err(|_| "Failed to get vc_issued_call_indexes".to_string())?
+			.map_err(|_| "Failed to get metadata".to_string())?;
+
+		let key = maybe_key.ok_or_else(|| "Invalid aes key".to_string())?;
 		let call = OpaqueCall::from_tuple(&(
 			call_index,
-			response.assertion_request.who,
-			response.assertion_request.assertion,
-			response.vc_index,
-			response.vc_hash,
-			H256::zero(),
+			who.clone(),
+			assertion,
+			res.vc_index,
+			res.vc_hash,
+			id_graph_hash,
+			req_ext_hash,
 		));
+
+		let mutated_id_graph = if should_create_id_graph { id_graph } else { Default::default() };
+
 		let res = RequestVCResult {
-			vc_index: response.vc_index,
-			vc_hash: response.vc_hash,
-			vc_payload: result,
+			vc_index: res.vc_index,
+			vc_hash: res.vc_hash,
+			vc_payload: aes_encrypt_default(&key, &res.vc_payload),
+			pre_mutated_id_graph: aes_encrypt_default(&key, &mutated_id_graph.encode()),
+			pre_id_graph_hash: id_graph_hash,
 		};
-		// This internally fetches nonce from a Mutex and then updates it thereby ensuring ordering
+
+		// submit TrustedCall::maybe_create_id_graph
+		let enclave_signer: AccountId = context
+			.enclave_signer
+			.get_enclave_account()
+			.map_err(|_| "Failed to get enclave signer".to_string())?;
+		let c = TrustedCall::maybe_create_id_graph(enclave_signer.into(), who);
+		context
+			.submit_trusted_call(&request.shard, None, &c)
```

Comment:
Have you tried multiple requests that try to create idgraph? This when run in parallel in threadpools will likely fail due to nonce. The nonce calculation takes the existing nonce and the pending trusted calls, so it ends up having the same nonce when run in parallel. Maybe we could have a unit test or `ts-test` for this case.

Reply:
This is a good point.

I think either adding mutex or redirecting them to single-thread execution (e.g. using `mpsc`) to tackle this, wdyt?

Reply:
Yeah either of it works, I think  I've tackled it in `stf-thread` to redirect for single-threaded execution - https://github.com/litentry/litentry-parachain/pull/2159

Reply:
Fixed that - asking for a re-review